src/MycilaESPConnect.cpp

```diff
@@ -376,15 +376,41 @@ void Mycila::ESPConnect::loop() {
     _disableCaptivePortal();
   }
 
-  if (_state == Mycila::ESPConnect::State::PORTAL_COMPLETE || _state == Mycila::ESPConnect::State::PORTAL_TIMEOUT) {
+  if (_state == Mycila::ESPConnect::State::PORTAL_TIMEOUT) {
+    LOGW(TAG, "Portal timeout, restarting ESP...");
     _stopAP();
+    ESP.restart();
+  }
+
+  if (_state == Mycila::ESPConnect::State::PORTAL_COMPLETE) {
     if (_autoRestart) {
-      LOGW(TAG, "Auto Restart of ESP...");
-      ESP.restart();
-    } else
+      if (_restartRequestTime != 0) {
+        // init timeout 
+        LOGW(TAG, "Restart in %d ms...", _restartRequestTime);
+        _restartDelay = millis();
+        if(_preRestartCallback)
+          _preRestartCallback();
+      } else if (millis() - _restartDelay >= _restartRequestTime) {
+        // delay is over restart
+        if (_restartCallback)
+          _restartCallback();
+        LOGW(TAG, "Auto Restart of ESP...");
+        _stopAP();
+        
+        // We stop the AP and restart to be done
+        ESP.restart();
+      } else if (_restartRequestTime == 0) {
+        // if the delay is set to zero, we skip the restart
+        // and we don't want to restart again
+        LOGW(TAG, "Restart delay is set to zero, skipping restart."); 
+        _restartDelay = 0;
+        _autoRestart = false;
+      }
```

Comment:
more something like this:

```suggestion
    if (_restartRequestTime == 0) {
      // first time: record the time of the restart request and trigger the callback
      _restartRequestTime = millis();
       _stopAP();
    }
    if (millis() - _restartRequestTime >= _restartDelay) {
        LOGW(TAG, "Auto Restart of ESP...");
        ESP.restart();
    }
```

Reply:
when I do _stopAP(); too early, the user doesn't receive the last request and I'll need to delay _stopAP(); to test the new connection

Reply:
In that case, use `setRestartDelay(1000)` for example.
also, _restartDelay should be defaulted to 1000.

There is no need to add a blocking delay(): the idea is to make the loop go in the code to check for the condition, when the delay is passed, trigger the restart.